### ISSUE TITLE
Tracking: WebAssembly support for iroh

### ISSUE THREAD

**Issue by matheus23** [n0-computer/iroh#2799]
We're working on Wasm support so you can use iroh-net in rust applications in browsers. See also [the accompanying blogpost](https://www.iroh.computer/blog/iroh-and-the-web).

Here's some prior issues with requests for this:
- https://github.com/n0-computer/iroh/issues/1803
- https://github.com/n0-computer/iroh/issues/2671

This is a **non-exhaustive** list of things that need to land for Wasm support:
- https://github.com/n0-computer/iroh/pull/2806
- https://github.com/n0-computer/iroh/issues/1218
- quinn upstreaming (not strictly necessary, since we're on a fork, but *definitely* helpful):
  - https://github.com/quinn-rs/quinn/pull/1996
  - https://github.com/quinn-rs/quinn/pull/2037
- https://github.com/n0-computer/iroh/issues/2979 (not strictly necessary in full, but *some* indirection for `tokio::spawn` *is* necessary)

**Comment by amiyatulu** [comment]
Thanks for the blogspot, and about relay servers. Will [Ockam](https://www.ockam.io/) be helpful in [end to end encryption](https://youtu.be/jKvIeQFZKLc?si=Bzaf0xcHF8uAJkbQ)?

**Comment by matheus23** [comment]
> Thanks for the blogspot, and about relay servers. Will [Ockam](https://www.ockam.io/) be helpful in [end to end encryption](https://youtu.be/jKvIeQFZKLc?si=Bzaf0xcHF8uAJkbQ)?

The traffic that's relayed is already encrypted end-to-end, because we're using QUIC :) (Yes, we're compiling a whole QUIC network stack to Wasm and run it in the browser :upside_down_face: ).



> Also can I use any relay server to connect, other wise it will lead to centralization (e.g. storing relay server addresses in dht)

Yes, relay servers are identified by URL. As a node, you can decide to use any "home" relay server you want.
By default, number 0 runs three: one in Asia, one in Europe and one in the US (at the moment).
But you can simply run one on your own via the binary behind the `relay` feature in `iroh-net` + DNS and some TLS config.
And if there are any relay servers you want to use from iroh, you configure them via the [`RelayMap::from_url`](https://docs.rs/iroh-net/latest/iroh_net/relay/struct.RelayMap.html#method.from_url) (or configure a whole set using `RelayMap::from_nodes`).

**Comment by matheus23** [comment]
For anyone who's following this issue: We've got preliminary iroh support for browsers via WebAssembly.
During this preliminary support phase, you'll need to depend on a branch to be able to use it, but we'll release it properly soon.

To try it out, add this git dependency to your `Crates.toml`:
```toml
iroh = { git = "https://github.com/n0-computer/iroh.git", branch = "matheus23/iroh-browser", default-features = false }
```

Running `cargo build --target=wasm32-unknown-unknown` should now just work, assuming the rest of your project is set up to support browsers, too.

Some caveats:
- WebAssembly support won't be compatible with the features `metrics`, `test-utils`, `discovery-local-network`, nor `discovery-pkarr-dht`, unfortunately.
- Browser support specifically hooks in to [`wasm-bindgen`](https://rustwasm.github.io/docs/wasm-bindgen/)s bindings generation, so using iroh in browser requires setting up wasm-bindgen. We also leverage [`wasm-bindgen-futures`](https://docs.rs/wasm-bindgen-futures), so async iroh code will use javascript promises on the main browser thread.
- Using iroh from browsers will be similar to using iroh in a "relay only" mode. When you connect to another node, the traffic *will* be end-to-end encrypted for the node you connect to in particular, however, it won't try to hole-punch or try to establish a direct connection, as that's not possible in browsers without deeply integrating with WebRTC. Instead, traffic is relayed via a WebSocket connections to relays.

Remember, this is specifically support for iroh *in browsers*, which we expect to be the biggest use case for compiling iroh to WebAssembly. We don't support other related things such as compiling to `wasm32-unknown-unknown` without `wasm-bindgen` nor compiling to `wasm32-wasi` or similar.

**Comment by satvikpendem** [comment]
> WebAssembly support won't be compatible with the features metrics, test-utils, discovery-local-network, nor discovery-pkarr-dht, unfortunately.

Is this just for the current alpha release or it won't be compatible at all with WASM in the future as well?

**Comment by matheus23** [comment]
> Is this just for the current alpha release or it won't be compatible at all with WASM in the future as well?

`metrics` and `test-utils` feature support might come in the future.

However, `discovery-local-network` relies on `mDNS`, which is simply unavailable in *browser* sandboxes today.
Similarly, `discovery-pkarr-dht` relies on UDP sockets being available to participate in the mainline DHT, which are also unavailable in browsers.
So no, we cannot provide support for these features in browsers.

You mentioned "with Wasm", though, so, are you referring to using Wasm in environments *outside* the browser?
I do think there's a path forward eventually with targeting wasm-wasi. We haven't looked into that yet, but if you have a use-case where you can deploy wasm-wasi, but cannot deploy native libraries/binaries, then this might be possible eventually.

**Comment by TotalKrill** [comment]
> Instead, traffic is relayed via a WebSocket connections to relays.

Is this always the case? My questions is because I am interested in solutions where regular servers also participate, and opening a direct websocket connection to a server from a browser client is very possible. But the other way is not as easy. 

By regular server, I mean applications running on a PC (windows, linux, mac)

**Comment by matheus23** [comment]
Yes. In the current iroh browser alpha, iroh will always maintain a websocket connection to its home relay.

Keep in mind though: iroh in the browser can totally talk to iroh running as a regular native application! They are 100% compatible.

**Comment by TotalKrill** [comment]
Thank you for that answer! I hope that when it leaves alpha, it will be able to establish websockets connections directly between clients if possible!

I have this idea where I would like browserbased clients to scuttlebut information to PC nodes that are locked inside internet less networks that the clients can join/leave. But I guess I would be limited by local-discovery not being available before direct websocket connections becomes an issue anyway

**Comment by satvikpendem** [comment]
> You mentioned "with Wasm", though, so, are you referring to using Wasm in environments outside the browser?

Thanks, I did mean just browsers as you mentioned the browser version uses WASM. Would be cool to see outside browsers too though.

**Comment by ethnh** [comment]
> * Using iroh from browsers will be similar to using iroh in a "relay only" mode. When you connect to another node, the traffic _will_ be end-to-end encrypted for the node you connect to in particular, however, it won't try to hole-punch or try to establish a direct connection, as that's not possible in browsers without deeply integrating with WebRTC. Instead, traffic is relayed via a WebSocket connections to relays.

is there any documentation on relays? 

Is there a way for WSS or WebTransport connections to do direct p2p? sad that webrtc isn't perfect https://github.com/w3c/webrtc-pc/issues/2553 https://github.com/w3c/webrtc-pc/pull/317#issuecomment-171037422 https://github.com/w3c/webrtc-pc/issues/230 https://github.com/arewedistributedyet/arewedistributedyet/issues/7

**Comment by matheus23** [comment]
> is there any documentation on relays?

Yeah: https://www.iroh.computer/docs/concepts/relay

> Is there a way for WSS or WebTransport connections to do direct p2p? sad that webrtc isn't perfect

Yes and no. In theory you can establish connections from a browser to another endpoint with a public IP address (and might in some network conditions even successfully hole-punch through stateful firewalls and NATs), but you can not connect two browser endpoints to each other directly.
There is no API for accepting incoming WebTransport connections in the browser, and even outgoing connections require to either having TLS set up on one end or use [`serverCertificateHashes`](https://developer.mozilla.org/en-US/docs/Web/API/WebTransport/WebTransport#servercertificatehashes).

That said, `serverCertificateHashes` might not be a tool to rely on long-term: https://github.com/w3c/webtransport/issues/623

**Comment by arlyon** [comment]
With regards to local network detection, I have rolled my own P2P communication thing for a project I've been working on (litehouse.arlyon.dev) on top of webrtc and currently use a custom signaling server that lets you discover foreign nodes that share your IP address as resolved by the signaling node. I realise that custom signaling servers are off the table for iroh but would it be feasible for iroh to offer a similar discovery system to complement mdns?

**Comment by matheus23** [comment]
No, currently `sendme` is a CLI, so you'd have to write a front-end for sendme.

Another roadblock is also that sendme uses `iroh-blobs` and that hasn't gotten Wasm support yet.

**Comment by DougAnderson444** [comment]
Is there a tracking issue for `iroh-blobs` to target `wasm32`?

**Comment by matheus23** [comment]
[There is now](https://github.com/n0-computer/iroh-blobs/issues/90) :)

**Comment by teohhanhui** [comment]
Supporting WASI would be very nice, especially `wasm32-wasip2` or `wasm32-wasip3`.

Use cases would be platforms such as:
* [Spin](https://spinframework.dev/)
* [wasmCloud](https://wasmcloud.com/)

See also:
* https://arxiv.org/html/2509.09400v1

**Comment by matheus23** [comment]
Yeah, at number 0 we agree.
The strategy for wasm32-wasip2/3 would be very different from wasm32-unknown-unknown/wasm-bindgen though.
Instead of swapping out tokio with another runtime (wasm-bindgen-futures/"the browser" in that case), removing any code that uses UDP sockets, and replacing anything that uses HTTPS over TCP with browser's `fetch`, we'd instead try to compile most of the stack to wasm32-wasip2/3.

So this means we'd be dependent on tokio to work under that platform and for it to support UDP/TCP sockets: https://github.com/tokio-rs/tokio/issues/4827
And we'd *might* have to look at our net-tools again and what the status of listing interfaces in WASI looks like.
